renter/rpc.go

```diff
@@ -308,16 +308,21 @@ func (s *Session) FormContract(renterKey types.PrivateKey, renterFunds, hostColl
 		hostInputs[i].Signatures = hostSigs.SiacoinInputSignatures[i]
 	}
 
+	// the transaction should now be valid
+	if err := vc.ValidateTransaction(txn); err != nil {
+		return rhp.Contract{}, types.Transaction{}, fmt.Errorf("renewal transaction is invalid: %w", err)
+	}
+
 	return rhp.Contract{
 		ID:       txn.FileContractID(0),
 		Revision: txn.FileContracts[0],
 	}, txn, nil
 }
 
-// RenewContract renews an existing contract, modifying its funding and/or
-// duration. Funds remaining in the old contract will be rolled over into the
-// new contract.
-func (s *Session) RenewContract(renterKey types.PrivateKey, contract types.FileContractRevision, renterFunds, validHostPayout, missedHostPayout types.Currency, endHeight uint64) (rhp.Contract, types.Transaction, error) {
+// RenewContract renews an existing contract, adding additional funds,
+// collateral and/or duration. Funds remaining in the old contract will be
+// rolled over into the new contract.
+func (s *Session) RenewContract(renterKey types.PrivateKey, contract types.FileContractRevision, additionalRenterFunds, additionalHostCollateral types.Currency, endHeight uint64) (rhp.Contract, types.Transaction, error) {
```

Comment:
I'd prefer to leave this more generic:
```go
func (s *Session) RenewContract(renterKey types.PrivateKey, contract types.FileContractRevision, renterFunds, validHostCollateral, missedHostCollateral types.Currency, endHeight uint64)
```
I'm just not super confident that the renewal math is correct for all cases. Better to punt those calculations up the stack; then we can implement wrappers like `ExtendContract` (increase endheight)  and `RefillContract` (keep same endheight) with more straightforward math, and we won't have to keep tweaking this API.

Reply:
There are only three possible scenarios: the renter is extending the length of the contract, adding funds, or both.

Correct me if I'm wrong, but I don't believe we can use that method signature because it doesn't capture the portion of the renter's funds that should be allocated to the valid host output to pay for the storage extension. Since the host has to validate the renewal pays them enough, we have to get the math correct on that end regardless.

Reply:
In principle, you could have a renewal with a shorter duration and less funds! The only thing that distinguishes a renewal from a new contract is that the host is already storing some data (and thus, should be compensated for storing that data, and should be liable for losing that data). If the host greatly lowers their prices after the initial contract formation, then you could end up paying less for a renewal even if it extends the duration. So I don't like calling these parameters `additional`, because the amount of funds/collateral isn't necessarily going to increase.

I guess the host parameters should be `validHostPayout` and `missedHostPayout` (since they aren't *purely* collateral), but in any case we know that only three currency values are needed (since the renter payout is the same for both valid and missed).

Reply:
Renewals for shorter durations don’t affect the existing storage cost though. The renter has already payed for the duration at the time of upload as part of the existing contract. I’m not sure it’s worth it to change the channel so we can provide an accurate refund 🤔

Currently, the only part the renter *can* change is how much additional funding they want to add to the contract and it’s expiration. Then we need to determine how much additional storage time they’re using and pay the host for that. So renter funds is really two variables, renter storage extension cost and renter unused funds

Reply:
hmm. I guess it simplifies things to just ban renewals with a lower end height. In that case, we should use an `extension` parameter rather than `endHeight`.

Maybe things would be clearer if we defined `FileContract` like so:
```go
type FileContract struct {
    // other fields omitted

    RenterOutput   types.SiacoinOutput
    HostOutput     types.SiacoinOutput

    HostCollateralAvailable types.Currency
    HostCollateralRisked    types.Currency
}
```
We initialize the contract with `RenterOutput = renterFunds`, `HostOutput = settings.ContractFee`, `HostCollateralAvailable = hostCollateral`, and `HostCollateralRisked = 0`. The renter pays for `RenterOutput + HostOutput`, and the host pays for `HostCollateralAvailable`. To revise the contract, we move value from `RenterOutput` to `HostOutput`; if we're uploading, we also move value from `HostCollateralAvailable` to `HostCollateralRisked`. If the contract resolves valid, then we disburse `RenterOutput` and `HostOutput + HostCollateralAvailable + HostCollateralRisked`; but if it resolves invalid, then we only disburse `HostOutput + HostCollateralAvailable`.

So: what does renewal look like with this design?

The renter can freely calculate new values for `RenterOutput` and `HostCollateralAvailable`, based on how much it wants to spend on the contract and how much collateral it wants the host to make available. The initial `HostOutput` should be `settings.ContractFee`, as before, and `HostCollateralRisked` should be recalculated as `filesize * settings.Collateral`. The renter funds the transaction with `RenterOutput + HostOutput - renterRollover`, and the host funds `HostCollateralAvailable + HostCollateralRisked - hostRollover`.

There's just one annoyance, though. When we renew a contract, we're also finalizing the old contract. In the old contract, the renter was paying the host to store data for X blocks -- but now, we're renewing after only Y blocks, so we're overpaying the host!

I see two ways of dealing with this. First, we can correct the overpayment in the final revision; to do this, move `filesize * storage cost * remaining duration` from the host back to the renter. Second, we can require the host to rollover *at least this amount* into the new contract. Both have tradeoffs; open to suggestions here.

Reply:
> If the contract resolves valid, then we disburse RenterOutput and HostOutput + HostCollateralAvailable + HostCollateralRisked; but if it resolves invalid, then we only disburse HostOutput + HostCollateralAvailable.

Shouldn't `RenterOutput` always be disbursed since it's the *unspent* renter funds? Would we need a different value to track the spent and burnt funds?

Hmm, does this take into account the renter's burnt allowance?

> When we renew a contract, we're also finalizing the old contract. In the old contract, the renter was paying the host to store data for X blocks -- but now, we're renewing after only Y blocks, so we're overpaying the host!

If the renter only pays for the extension, it's not overpaying the host. The renter only pays the difference between the original and the new end. The host isn't risking as much collateral though since the risk would be the extension duration, not the extension + remaining.

Reply:
yeah, I omitted it for brevity, but the renter gets their output regardless. Not sure what you mean about the burn field?

By "overpayment," I'm referring to how the finalization releases outputs immediately. If the renter paid the host to store for 10 blocks, but renews after 5 blocks, then the finalization output will still pay the host as though they stored for 10 blocks.